### PR TITLE
fix(RHINENG-5391): Fix hosts groups with wrong value in DB

### PR DESCRIPTION
**Pull request by jpramos123**
# Overview

This PR is being created to address [RHINENG-5391](https://issues.redhat.com/browse/RHINENG-5391).

This code is meant to fix the hosts that contains a wrong value for the groups field.
Today some hosts have the value of groups field equal to "{}", which should not be possible.
It is validated that the Host MQ does not accept this value during the host insertion anymore.

The solution is to perform an simple update call to the hosts table, finding the hosts WHERE groups = "{}" and update it to groups = "[]", which is current default state for new hosts that are not associated to any inventory group.

Produced message:

```
[2024-08-12 17:51:33,903] [148030] [139710726680640] [inventory.migrations.11cd38415793_fix_hosts_groups_with_wrong_value_in_db_py] [INFO] Matched rows for UPDATE: 5
```

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [x] Descriptive comments provided in complex code blocks
- [ ] Include raw query examples in the PR description, if adding/modifying SQL query
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

**Comment by fstavela** [comment]
The pipeline says that pre-commit is failing:
```
10:04:05 Reorder python imports...................................................Passed
10:04:05 pyupgrade................................................................Passed
10:04:09 black....................................................................Passed
10:04:09 trim trailing whitespace.................................................Passed
10:04:09 fix end of files.........................................................Passed
10:04:09 check yaml...............................................................Passed
10:04:09 debug statements (python)................................................Passed
10:04:11 flake8...................................................................Passed
10:04:11 swagger-validation.......................................................Passed
10:04:11 swagger-validation.......................................................Passed
10:04:11 redoc filter.............................................................Failed
10:04:11 - hook id: redocly-filter
10:04:11 - exit code: 1
10:04:11 
10:04:11 Executable `redocly` not found
10:04:11 
10:04:11 redoc validate...........................................................Failed
10:04:11 - hook id: redocly-validate
10:04:11 - exit code: 1
10:04:11 
10:04:11 Executable `redocly` not found
```

File: host_update_host_groups.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+def _init_db(config):
+    engine = create_engine(config.db_uri)
+    return sessionmaker(bind=engine)
+
+
+with application.app.app_context():
+    config = _init_config()
+    Session = _init_db(config)
+    session = Session()
+    query = update(Host).where(Host.groups == "{}").values(groups="[]").returning(Host.id)
```
Comment (thearifismail) [inline host_update_host_groups.py:55]: Should use CHUNK_SIZE for updating hosts like when hosts are deleted.  The reason is, as of now, the Prod DB has a total of `44,964` hosts with `groups` value of `{}`; out of which `44,342` hosts are active, i.e. stale_timestamp is today or in future.

Comment (RonnyPfannschmidt) [reply host_update_host_groups.py:55]: dont return host ids there, count them before -

Comment (RonnyPfannschmidt) [reply host_update_host_groups.py:55]: dont return host ids there, count them before -

**Review by RonnyPfannschmidt — CHANGES_REQUESTED** [review]
this really should be a database migration

**Comment by thearifismail** [comment]
@joao have you tested the migration script by running `make upgrade_db` targeting a DB container` with bad host group values that should be fixed?

**Review by thearifismail — COMMENTED** [review]
please see the comment

**Comment by jpramos123** [comment]
> @joao have you tested the migration script by running `make upgrade_db` targeting a DB container` with bad host group values that should be fixed?

Yes! Definitely, and it works just fine!

You can simulate it locally as well!

Thanks

File: migrations/versions/11cd38415793_fix_hosts_groups_with_wrong_value_in_db.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+class Host(Base):
+    __tablename__ = "hosts"
+    id = Column(UUID, primary_key=True)
+    groups = Column(JSONB(none_as_null=True))
+
+
+def _fix_host_groups(logger_, host_query):
+    filtered_query = host_query.filter(Host.groups == "{}")
+    matched_rows = filtered_query.update({Host.groups: "[]"})
```
Comment (roliveri) [inline migrations/versions/11cd38415793_fix_hosts_groups_with_wrong_value_in_db.py:58]: Just checking. This returns the actual number of rows that were updated?

**Review by kruai — APPROVED** [review]
Seems to work, LGTM!

File: migrations/versions/11cd38415793_fix_hosts_groups_with_wrong_value_in_db.py (hunk 2 above)
Comment (jpramos123) [reply migrations/versions/11cd38415793_fix_hosts_groups_with_wrong_value_in_db.py:58]: @roliveri Correct! This is the produced message:

```
[2024-08-12 17:51:33,903] [148030] [139710726680640] [inventory.migrations.11cd38415793_fix_hosts_groups_with_wrong_value_in_db_py] [INFO] Matched rows for UPDATE: 5
```

**Pull request merged**